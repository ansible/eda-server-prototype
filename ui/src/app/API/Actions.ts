import { defaultSettings } from '@app/shared/pagination';
import { getAxiosInstance } from '@app/API/baseApi';
import { AxiosResponse } from 'axios';

const actionsRulesEndpoint = '/api/audit/rules_fired';
const actionsHostsEndpoint = '/api/audit/hosts_changed';
const actionsRuleEndpoint = '/api/audit/rule';

export const listActionsRules = (pagination = defaultSettings): Promise<AxiosResponse> =>
  getAxiosInstance().get(actionsRulesEndpoint);

export const listActionsHosts = (pagination = defaultSettings): Promise<AxiosResponse> =>
  getAxiosInstance().get(actionsHostsEndpoint);

export const fetchActionsRuleDetails = (ruleId: string | number, pagination = defaultSettings): Promise<AxiosResponse> =>
  getAxiosInstance().get(`${actionsRuleEndpoint}/${ruleId}/details`);

export const listActionsRuleJobs = (ruleId: string | number, pagination = defaultSettings): Promise<AxiosResponse> =>
  getAxiosInstance().get(`${actionsRuleEndpoint}/${ruleId}/jobs`);

export const listActionsRuleEvents = (ruleId: string | number, pagination = defaultSettings): Promise<AxiosResponse> =>
  getAxiosInstance().get(`${actionsRuleEndpoint}/${ruleId}/events`);

export const listActionsRuleHosts = (ruleId: string | number, pagination = defaultSettings): Promise<AxiosResponse> =>
  getAxiosInstance().get(`${actionsRuleEndpoint}/${ruleId}/hosts`);
