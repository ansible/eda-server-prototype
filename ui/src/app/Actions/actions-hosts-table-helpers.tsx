import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';
import { Text, TextVariants } from '@patternfly/react-core';

export const createRows = (data) =>
  data.map(({ host, rule, ruleset, fired_date }) => ({
    cells: [
      host,
      <Fragment key={`[actions-host-${rule?.id}`}>
        <Link
          to={{
            pathname: `/rule/${rule?.id}`,
          }}
        >
          {rule?.name || rule?.id}
        </Link>
      </Fragment>,
      <Fragment key={`[actions-ruleset-${ruleset?.id}`}>
        <Link
          to={{
            pathname: `/ruleset/${ruleset?.id}`,
          }}
        >
          {ruleset?.name || ruleset?.id}
        </Link>
      </Fragment>,
      <Fragment key={`[actions-last_fired-${rule?.name}`}>
        <Text component={TextVariants.small}>
          {new Intl.DateTimeFormat('en-US', { dateStyle: 'short', timeStyle: 'long' }).format(
            new Date(fired_date || 0)
          )}
        </Text>
      </Fragment>,
    ],
  }));
