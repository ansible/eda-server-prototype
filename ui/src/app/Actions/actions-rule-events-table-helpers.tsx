import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';
import {Text, TextVariants} from "@patternfly/react-core";

export const createRows = (data) =>
  data.map(({ id, name, increment_counter, type, timestamp }) => ({
    id,
    cells: [
      <Fragment key={`[event-${id}`}>
        <Link
          to={{
            pathname: `/event/${id}`,
          }}
        >
          {name || id}
        </Link>
      </Fragment>,
      increment_counter,
      type,
      <Fragment key={`[actions-rule-${id}-event`}>
        <Text component={TextVariants.small}>
          {new Intl.DateTimeFormat('en-US', { dateStyle: 'short', timeStyle: 'long' }).format(
            new Date(timestamp || 0)
          )}
        </Text>
      </Fragment>,
    ],
  }));
