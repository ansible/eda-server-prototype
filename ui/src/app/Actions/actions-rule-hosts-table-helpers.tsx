import React, { Fragment } from 'react';
import { statusLabel } from '@app/utils/utils';

export const createRows = (data, intl) => {
  return data.map(({ name, status }) => ({
    cells: [name, <Fragment key={`[actions-rule-hosts-${name}`}>{statusLabel({ text: status, intl: intl })}</Fragment>],
  }));
};
