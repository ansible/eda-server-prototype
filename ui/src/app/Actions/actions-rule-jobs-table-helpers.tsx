import React, { Fragment } from 'react';
import { Link } from 'react-router-dom';
import { statusLabel } from '@app/utils/utils';

export const createRows = (data, intl) =>
  data.map(({ id, name, status, last_fired_date }) => ({
    id,
    cells: [
      <Fragment key={`[job-${id}`}>
        <Link
          to={{
            pathname: `/job/${id}`,
          }}
        >
          {name || id}
        </Link>
      </Fragment>,
      <Fragment key={`[actions-rule-jobs-${name}`}>{statusLabel({ text: status, intl: intl })}</Fragment>,
      last_fired_date,
    ],
  }));
