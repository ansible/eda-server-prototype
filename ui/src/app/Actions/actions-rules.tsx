import { Button, PageSection } from '@patternfly/react-core';
import React, { useEffect, useReducer, useState } from 'react';
import { TableToolbarView } from '@app/shared/table-toolbar-view';
import sharedMessages from '../messages/shared.messages';
import TableEmptyState from '@app/shared/table-empty-state';
import { useIntl } from 'react-intl';
import { defaultSettings } from '@app/shared/pagination';
import { createRows } from './actions-rules-table-helpers';
import { CubesIcon } from '@patternfly/react-icons';
import { renderActionsTabs } from '@app/Actions/Actions';
import { ActionsRuleType } from '@app/Actions/Actions';
import { listActionsRules } from '@app/API/Actions';

const columns = (intl) => [
  {
    title: intl.formatMessage(sharedMessages.name),
  },
  {
    title: intl.formatMessage(sharedMessages.job),
  },
  {
    title: intl.formatMessage(sharedMessages.status),
  },
  {
    title: intl.formatMessage(sharedMessages.ruleset),
  },
  {
    title: intl.formatMessage(sharedMessages.lastFiredDate),
  },
];

const initialState = (filterValue = '') => ({
  filterValue,
  isFetching: false,
  isFiltering: false,
  rows: [],
});

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const actionsRulesListState = (state, action) => {
  switch (action.type) {
    case 'setRows':
      return {
        ...state,
        rows: action.payload,
      };
    case 'setFetching':
      return {
        ...state,
        isFetching: action.payload,
      };
    case 'setFilterValue':
      return { ...state, filterValue: action.payload };
    case 'setFilteringFlag':
      return {
        ...state,
        isFiltering: action.payload,
      };
    case 'clearFilters':
      return { ...state, filterValue: '', isFetching: true };
    default:
      return state;
  }
};

const ActionsRules: React.FunctionComponent = () => {
  const [limit, setLimit] = useState(defaultSettings.limit);
  const [offset, setOffset] = useState(1);
  const [actionsRules, setActionsRules] = useState<ActionsRuleType[]>([]);

  const meta = { count: actionsRules?.length || 0, limit, offset };

  const [{ filterValue, isFetching, isFiltering, rows }, stateDispatch] = useReducer(
    actionsRulesListState,
    initialState()
  );

  const intl = useIntl();
  const updateActionsRules = (pagination) => {
    stateDispatch({ type: 'setFetching', payload: true });
    return listActionsRules(pagination)
      .then((data) => {
        setActionsRules(data?.data);
        return stateDispatch({ type: 'setFetching', payload: false });
      })
      .catch(() => stateDispatch({ type: 'setFetching', payload: false }));
  };

  useEffect(() => {
    updateActionsRules(defaultSettings);
  }, []);

  useEffect(() => {
    stateDispatch({ type: 'setRows', payload: createRows(actionsRules, intl) });
  }, [actionsRules]);

  const clearFilters = () => {
    stateDispatch({ type: 'clearFilters' });
    return updateActionsRules(meta);
  };

  const handleFilterChange = (value) => {
    !value || value === '' ? clearFilters() : stateDispatch({ type: 'setFilterValue', payload: value });
  };

  return (
    <PageSection page-type={'activation-details'} id={'activation-details'}>
      {renderActionsTabs(intl)}
      <TableToolbarView
        ouiaId={'actions-rules-table'}
        rows={rows}
        setLimit={setLimit}
        setOffset={setOffset}
        columns={columns(intl)}
        fetchData={updateActionsRules}
        pagination={defaultSettings}
        plural={intl.formatMessage(sharedMessages.rules)}
        singular={intl.formatMessage(sharedMessages.rule)}
        isLoading={isFetching || isFiltering}
        onFilterChange={handleFilterChange}
        renderEmptyState={() => (
          <TableEmptyState
            title={intl.formatMessage(sharedMessages.norulesrecentlyfired)}
            Icon={CubesIcon}
            PrimaryAction={() =>
              filterValue !== '' ? (
                <Button onClick={() => clearFilters()} variant="link">
                  {intl.formatMessage(sharedMessages.clearAllFilters)}
                </Button>
              ) : null
            }
            description={
              filterValue === ''
                ? intl.formatMessage(sharedMessages.norulesrecentlyfired)
                : intl.formatMessage(sharedMessages.clearAllFiltersDescription)
            }
          />
        )}
      />
    </PageSection>
  );
};

export { ActionsRules };
