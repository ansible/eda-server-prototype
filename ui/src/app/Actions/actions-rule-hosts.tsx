import { Button, PageSection } from '@patternfly/react-core';
import React, { useEffect, useReducer, useState } from 'react';
import { TableToolbarView } from '@app/shared/table-toolbar-view';
import sharedMessages from '../messages/shared.messages';
import TableEmptyState from '@app/shared/table-empty-state';
import { useIntl } from 'react-intl';
import { defaultSettings } from '@app/shared/pagination';
import { createRows } from './actions-rule-hosts-table-helpers';
import { CubesIcon } from '@patternfly/react-icons';
import {listActionsRuleHosts} from '@app/API/Actions';
import { RuleType } from '@app/Rules/Rules';
import { renderActionsRuleTabs } from '@app/Actions/ActionsRule';

export interface HostType {
  name: string;
  status?: string;
}

const columns = (intl) => [
  {
    title: intl.formatMessage(sharedMessages.name),
  },
  {
    title: intl.formatMessage(sharedMessages.status),
  },
];

const initialState = (filterValue = '') => ({
  filterValue,
  isFetching: false,
  isFiltering: false,
  rows: [],
});

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const hostsListState = (state, action) => {
  switch (action.type) {
    case 'setRows':
      return {
        ...state,
        rows: action.payload,
      };
    case 'setFetching':
      return {
        ...state,
        isFetching: action.payload,
      };
    case 'setFilterValue':
      return { ...state, filterValue: action.payload };
    case 'setFilteringFlag':
      return {
        ...state,
        isFiltering: action.payload,
      };
    case 'clearFilters':
      return { ...state, filterValue: '', isFetching: true };
    default:
      return state;
  }
};

const ActionsRuleHosts: React.FunctionComponent<{ rule: RuleType }> = ({ rule }) => {
  const [limit, setLimit] = useState(defaultSettings.limit);
  const [offset, setOffset] = useState(1);
  const [hosts, setHosts] = useState<HostType[]>([]);

  const meta = { count: hosts?.length || 0, limit, offset };
  const [{ filterValue, isFetching, isFiltering, rows }, stateDispatch] = useReducer(hostsListState, initialState());

  const intl = useIntl();
  const updateHosts = (pagination) => {
    stateDispatch({ type: 'setFetching', payload: true });
    return listActionsRuleHosts(rule.id, pagination)
      .then((data) => {
        setHosts(data?.data);
        return stateDispatch({ type: 'setFetching', payload: false });
      })
      .catch(() => stateDispatch({ type: 'setFetching', payload: false }));
  };

  useEffect(() => {
    updateHosts(defaultSettings);
  }, []);

  useEffect(() => {
    stateDispatch({ type: 'setRows', payload: createRows(hosts, intl) });
  }, [hosts]);

  const clearFilters = () => {
    stateDispatch({ type: 'clearFilters' });
    return updateHosts(meta);
  };

  const handleFilterChange = (value) => {
    !value || value === '' ? clearFilters() : stateDispatch({ type: 'setFilterValue', payload: value });
  };
  return (
    <PageSection page-type={'actions-rule-hosts'} id={'actions-rule-hosts'}>
      {renderActionsRuleTabs(rule?.id, intl)}
      <TableToolbarView
        ouiaId={'actions-rule-hosts-table'}
        rows={rows}
        setLimit={setLimit}
        setOffset={setOffset}
        columns={columns(intl)}
        fetchData={updateHosts}
        pagination={defaultSettings}
        plural={intl.formatMessage(sharedMessages.hosts)}
        singular={intl.formatMessage(sharedMessages.host)}
        isLoading={isFetching || isFiltering}
        onFilterChange={handleFilterChange}
        renderEmptyState={() => (
          <TableEmptyState
            title={intl.formatMessage(sharedMessages.noactionsrulehosts)}
            Icon={CubesIcon}
            PrimaryAction={() =>
              filterValue !== '' ? (
                <Button onClick={() => clearFilters()} variant="link">
                  {intl.formatMessage(sharedMessages.clearAllFilters)}
                </Button>
              ) : null
            }
            description={
              filterValue === ''
                ? intl.formatMessage(sharedMessages.noactionsrulehosts)
                : intl.formatMessage(sharedMessages.clearAllFiltersDescription)
            }
          />
        )}
      />
    </PageSection>
  );
};

export { ActionsRuleHosts };
