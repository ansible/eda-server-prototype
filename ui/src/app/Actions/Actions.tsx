import React from 'react';
import { Level, LevelItem, Title } from '@patternfly/react-core';
import { Route, Switch, useLocation } from 'react-router-dom';
import { useIntl } from 'react-intl';
import AppTabs from '@app/shared/app-tabs';
import { TopToolbar } from '@app/shared/top-toolbar';
import { AnyObject, TabItemType } from '@app/shared/types/common-types';
import sharedMessages from '../messages/shared.messages';
import { ActionsRules } from '@app/Actions/actions-rules';
import { ActionsHosts } from '@app/Actions/actions-hosts';

export interface ActionsRuleType {
  id: string;
  name: string;
  job: string;
  job_name: string;
  status: string;
  ruleset: string;
  ruleset_name: string;
  fired_at?: string;
}

export interface ActionsHostType {
  rule: { id: string; name: string };
  job?: { id: string; name: string };
  ruleset?: { id: string; name: string };
  status: string;
  fired_date?: string;
}

const buildActionsTabs = (intl: AnyObject): TabItemType[] => [
  { eventKey: 0, title: intl.formatMessage(sharedMessages.actions_title), name: `/actions/rules` },
  {
    eventKey: 1,
    title: intl.formatMessage(sharedMessages.actions_hosts_title),
    name: `/actions/hosts`,
  },
];

export const renderActionsTabs = (intl) => {
  const actions_tabs = buildActionsTabs(intl);
  return <AppTabs tabItems={actions_tabs} defaultActive={0} />;
};

const Actions: React.FunctionComponent = () => {
  const intl = useIntl();

  return (
    <React.Fragment>
      <TopToolbar>
        <Level>
          <LevelItem>
            <Title headingLevel={'h2'}>{intl.formatMessage(sharedMessages.actions_view_title)}</Title>
          </LevelItem>
        </Level>
      </TopToolbar>
      <Switch>
        <Route exact path="/actions/hosts">
          <ActionsHosts />
        </Route>
        <Route path="/actions">
          <ActionsRules />
        </Route>
      </Switch>
    </React.Fragment>
  );
};

export { Actions };
