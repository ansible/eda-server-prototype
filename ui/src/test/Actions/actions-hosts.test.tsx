import * as React from 'react';
import { mount } from 'enzyme';
import { ActionsHosts } from '@app/Actions/actions-hosts';
import { MemoryRouter } from 'react-router';
import { act } from 'react-dom/test-utils';
import { IntlProvider } from 'react-intl';
import { Route } from 'react-router-dom';
import { mockApi } from '../__mocks__/baseApi';
import {ActionsRules} from "@app/Actions/actions-rules";

const ComponentWrapper = ({ children, initialEntries = ['/dashboard'] }) => (
  <IntlProvider locale="en">
    <MemoryRouter initialEntries={initialEntries} keyLength={0} key={'test'}>
      {children}
    </MemoryRouter>
  </IntlProvider>
);

describe('ActionsHosts', () => {
  beforeAll(() => {
    mockApi.reset();
  });

  it('should render the ActionsView component', async () => {
    mockApi.onGet(`/api/actions/rules_fired`).replyOnce(200, [
      { id: '1', name: 'Rule 1' },
      { id: '2', name: 'Rule 2' },
      { id: '3', name: 'Rule3 3' },
    ]);

    let wrapper;
    await act(async () => {
      wrapper = mount(
        <ComponentWrapper initialEntries={['/actions']}>
          <Route path="/actions/rules">
            <ActionsRules />
          </Route>
        </ComponentWrapper>
      );
    });
    await act(async () => {
      wrapper.update();
    });
    expect(wrapper).toMatchSnapshot();
  });
});
