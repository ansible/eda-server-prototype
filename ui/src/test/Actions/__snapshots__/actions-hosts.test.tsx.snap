// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActionsHosts should render the ActionsView component 1`] = `
<ComponentWrapper
  initialEntries={
    [
      "/actions",
    ]
  }
>
  <IntlProvider
    defaultFormats={{}}
    defaultLocale="en"
    fallbackOnEmptyString={true}
    formats={{}}
    locale="en"
    messages={{}}
    onError={[Function]}
    onWarn={[Function]}
    textComponent={Symbol(react.fragment)}
  >
    <MemoryRouter
      initialEntries={
        [
          "/actions",
        ]
      }
      key="test"
      keyLength={0}
    >
      <Router
        history={
          {
            "action": "POP",
            "block": [Function],
            "canGo": [Function],
            "createHref": [Function],
            "entries": [
              {
                "hash": "",
                "pathname": "/actions",
                "search": "",
                "state": undefined,
              },
            ],
            "go": [Function],
            "goBack": [Function],
            "goForward": [Function],
            "index": 0,
            "length": 1,
            "listen": [Function],
            "location": {
              "hash": "",
              "pathname": "/actions",
              "search": "",
              "state": undefined,
            },
            "push": [Function],
            "replace": [Function],
          }
        }
      >
        <Route
          path="/actions/rules"
        />
      </Router>
    </MemoryRouter>
  </IntlProvider>
</ComponentWrapper>
`;
